runtime/runtime-core/common/src/aws/smithy/kotlin/runtime/retries/impl/ExponentialBackoffWithJitter.kt

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+package aws.smithy.kotlin.runtime.retries.impl
+
+import aws.smithy.kotlin.runtime.retries.BackoffDelayer
+import kotlinx.coroutines.delay
+import kotlin.math.pow
+import kotlin.random.Random
+
+/**
+ * A [BackoffDelayer] that implements exponentially increasing delays and jitter (i.e., randomization of delay amount).
+ * This delayer calculates a maximum delay time from the initial delay amount, the scale factor, and the attempt number.
+ * It then randomly reduces that time down to something less based on the jitter configuration.
+ *
+ * For instance, a jitter
+ * configuration of 0.5 means that up to 50% of the max delay time could be reduced. A jitter configuration of 1.0 means
+ * that 100% of the max delay time could be reduced (potentially down to 0). A jitter configuration of 0.0 means jitter
+ * is disabled.
+ *
+ * @param options The configuration to use for this delayer.
+ */
+class ExponentialBackoffWithJitter(val options: ExponentialBackoffWithJitterOptions) : BackoffDelayer {
+    private val random = Random.Default
+
+    /**
+     * Delays for an appropriate amount of time after the given attempt number.
+     */
+    override suspend fun backoff(attempt: Int) {
+        require(attempt > 0) { "attempt was $attempt but must be greater than 0" }
+        val maxDelayMs = options.initialDelayMs * options.scaleFactor.pow(attempt - 1)
+        val jitterProportion = if (options.jitter > 0.0) random.nextDouble(options.jitter) else 0.0
+        val delayMs = maxDelayMs * (1.0 - jitterProportion)
+        delay(delayMs.toLong())
+    }
+}
+
+/**
+ * The configuration options for a [BackoffDelayer].
+ * @param initialDelayMs The initial amount of delay
+ */
+data class ExponentialBackoffWithJitterOptions(
+    val initialDelayMs: Int,
+    val scaleFactor: Double,
+    val jitter: Double,
+) {
+    init {
+        require(initialDelayMs > 0) { "initialDelayMs must be at least 0" }
+        require(scaleFactor >= 1.0) { "scaleFactor must be at least 1" }
+        require(jitter in 0.0..1.0) { "jitter must be between 0 and 1" }
+    }
+
+    companion object {
+        /**
+         * The default backoff configuration to use.
+         */
+        val Default = ExponentialBackoffWithJitterOptions(
+            initialDelayMs = 10, // start with 10ms
```

Comment:
comment

Thank you for the clarifying comment, very helpful

Reply:
My default posture is to document every public type and member, although I concede some members' very name and placement is the all documentation many will need. Do you think there's better documentation to be written here?

Reply:
No, I was just saying that these comments let me quickly understand how the parameters work.